spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java

```diff
@@ -250,7 +251,9 @@ public ScanBuilder newScanBuilder(CaseInsensitiveStringMap options) {
   @Override
   public WriteBuilder newWriteBuilder(LogicalWriteInfo info) {
     Preconditions.checkArgument(
-        snapshotId == null, "Cannot write to table at a specific snapshot: %s", snapshotId);
+        snapshotId == null || info.options().get(SparkWriteOptions.BRANCH) != null,
```

Comment:
@rdblue @aokolnychyi This goes back to the thread [above](https://github.com/apache/iceberg/pull/6651/files#r1105162329). After merging #6717 the SparkDataWriteTests fail at this validation because we are now setting a snapshotID to determine the schema. My solution here is to use the info.options().get(SparkWriteOptions.Branch) != null because if a write branch is explicitly passed in, it's expected that a snapshotID is defined.

Reply:
@amogh-jahagirdar I am good with this approach! Might need to think of a bigger refactoring in future. would like to hear what others say

Reply:
In this case I think we need to also check that the snapshot ID is the branch head that is written to?

Reply:
I think if branch is set, it snapshot ID will be set to branch head in code. But we can add an additional check too.

Reply:
@jackye1995 yeah that's a safer guarantee, was thinking if it's really necessary based on how we build the scan (I don't see how the snapshot can be anything but the branch head if a branch is set) but this is probably a case where we want to be defensive. So I'm inclined to change it, will wait for @aokolnychyi @rdblue thoughts